compiler/src/dotty/tools/dotc/core/SymDenotations.scala

```diff
@@ -1698,20 +1698,27 @@ object SymDenotations {
           && (!owner.is(Package)
              || sym.originDenotation.infoOrCompleter.match
                   case _: SymbolLoaders.SecondCompleter => sym.associatedFile == this.symbol.associatedFile
-                  case _ => false)
+                  case _: ClassInfo if recursively => sym.is(Package) || sym.associatedFile == this.symbol.associatedFile
+                  case other => false)
 
         if owner.isClass then
           for c <- owner.info.decls.toList if maybeChild(c) do
             c.ensureCompleted()
+            if recursively then completeChildrenIn(c, recursively)
       end completeChildrenIn
 
       if is(Sealed) || isAllOf(JavaEnumTrait) then
         if !is(ChildrenQueried) then
-          // Make sure all visible children are completed, so that
-          // they show up in Child annotations. A possible child is visible if it
-          // is defined in the same scope as `cls` or in the companion object of `cls`.
-          completeChildrenIn(owner)
-          completeChildrenIn(companionClass)
+          if !is(Enum) then // non-enum sealed class
+            // Make sure all children are completed, so that
+            // they show up in Child annotations.
+            completeChildrenIn(defn.RootClass, recursively = true)
```

Comment:
I spent some time trying to find something smarter than just starting at the `<root>` package - but neither `Symbol.topLevelClass` nor `Symbol.enclosingPackageClass` really helped. If some kind of method that returns packages/classes for a source file there already exists in the compiler, it is probably more performant than this (but I could not find anything like that).